### ISSUE TITLE
Upgrading from v1.12 - works from local dev website or unit test but will not work from Azure Cloud Server project

### ISSUE THREAD

**Issue by SubyBrue** [Azure/azure-documentdb-dotnet#540]
Trying to upgrade some legacy code from Azure.DocumentDB 1.12.0 to 1.22.0
Running on .Net Framework 4.6, Azure SDK 2,9

If I run a DocumentDB query from my unit tests, or within the web application that is the base project the query works fine without any issues.... 

However once I try to deploy it to an Azure Cloud Service Project either in my Dev, QA or even the Azure Emulator the call for the query returns with a "Bad Request - Value does not fall within the expected range." error.
Please note - I am connecting to the Dev CosmosDB account on my Azure account.

I've tried a variety of NuGet versions (1.15.0, 1.14.0, 1.13.4, 1.13.0) and they all have the same issue - runs outside of Azure but cannot run within the Dev, QA or emulator environment.

I've double checked the Azure Cloud Service project but don't see any settings that would be set for a specific version of Microsoft.Azure.DocumentDB or cause it to run fine locally and not within the Cloud Service Project.

**Comment by ealsur** [comment]
>If I run a DocumentDB query from my unit tests, or within the web application that is the base project the query works fine without any issues....
 However once I try to deploy it to an Azure Cloud Service Project either in my Dev, QA or even the Azure Emulator the call for the query returns with a "Bad Request - Value does not fall within the expected range." error.

When you work with the Emulator and the Web Application, it works fine then? 

If I understand correctly, the issue appears when running in a Web Role?

**Comment by SubyBrue** [comment]
No it only works when running a console based unit test or within a web application.

Once we attempt to use the Azure Emulator or push it to our Dev or Test environment it fails with that error.  Please note we are hitting the same Dev Cosmos endpoint in all situations.

Someone had also responded indicating it might be an issue returning the queries contents  .ToList() but that doesn’t make sense that it would only fail within the Cloud Service project but work fine otherwise

Sent from my iPhone

On Jul 10, 2018, at 6:15 AM, Matias Quaranta <notifications@github.com<mailto:notifications@github.com>> wrote:


If I run a DocumentDB query from my unit tests, or within the web application that is the base project the query works fine without any issues....
However once I try to deploy it to an Azure Cloud Service Project either in my Dev, QA or even the Azure Emulator the call for the query returns with a "Bad Request - Value does not fall within the expected range." error.

When you work with the Emulator and the Web Application, it works fine then?

If I understand correctly, the issue appears when running in a Web Role?

**Comment by ealsur** [comment]
Ok, when you mean Emulator you mean the Azure Cloud Emulator, not the Azure Cosmos DB Emulator.

Can you check in your WebRole project, if you have any dependency (Nuget package) with a conflicting / different version than the Web Application / Console Unit Test project? It could be that the WebRole is overriding a dependency or using a different dependency than your Web/Console project.

**Comment by SubyBrue** [comment]
Yes the Cloud Emulator for the SDK 2.9.....    We are hitting the Dev Cosmos DB normally....    Yes we have double checked that all NuGet versions are the same between the code layers.

One thing we did notice in trying to debug further is that when the Code runs in the Emulator or in Azure as the Cloud Service (where it doesn't work) - but works fine as a normal web project...   is that it tripps up when we try and deserialize the result with a .ToList() with v1.22   

Method: GetMetadata<T>()
        public static async Task<List<T>> GetMetaData<T>(string noSqlDbName, string queryString)
        {
            var client = GetDocumentDbClient(noSqlDbName);
            var collection = CosmosSingleton.Instance.GetDocColl(String.Empty, SearchUtility.SearchConfig.DocDbs.DatabaseName, SearchMgrEnty.Meta).UriInfo;

            var dq = client.CreateDocumentQuery<T>(collection, String.Format(queryString));

            return dq.ToList();
        }

I was able to extract this particular error during emulator runtime when we try to convert ToList().
error CS0030: Cannot convert type 'System.Linq.IQueryable<T>' to 'Microsoft.Azure.Documents.Linq.DocumentQuery<Microsoft.Azure.Documents.Document>'

But yet if I run in outside of the Azure Cloud Service it runs fine.   
Also if we attempt with v1.15 it just fails with a "Bad Request" Value Not In Range.

However when running v1.12 it runs fine in both Azure Cloud Service and as a Web Project.

**Comment by ealsur** [comment]
@SubyBrue I'm working on a repro of this locally and will share my findings.

**Comment by ealsur** [comment]
@SubyBrue I created a blank new Cloud Service project with DocumentDB SDK 1.22: https://github.com/ealsur/cloudservicedocdbsample

I did not have your Model (or actually the value of `<T>` you are using in the `CreateDocumentQuery` so I created [one based on your query](https://github.com/ealsur/cloudservicedocdbsample/blob/master/WebRole1/Models/MetaData.cs)). 

It is a Cloud Service with 1 WebRole containing in simple ASP.NET MVC App.

The sample connects to the [Azure Cosmos DB Emulator](https://docs.microsoft.com/en-us/azure/cosmos-db/local-emulator) (you don't need to use a live account), it will [create a sample collection](https://github.com/ealsur/cloudservicedocdbsample/blob/master/WebRole1/WebRole.cs#L17) and some documents `OnStart` just to have something to query on.

Then on the main Controller, when loading the default View, it will [run the query you suggested](https://github.com/ealsur/cloudservicedocdbsample/blob/master/WebRole1/Controllers/HomeController.cs#L17) and return the results to the client.

Running it in the local Cloud Emulator 2.9 it works correctly and does not throw any error.

If you could share what is `<T>`, or what model are you passing, it might shed some light on your issue since it seems to be related to that.

**Comment by SubyBrue** [comment]
@ealsur    I'll look at your blank Cloud Service project tomorrow.        
We are on .Net Framework 4.6 (Not Core or Standard).    Our Cloud Service has 2 WebRoles in it  which has a webproject that exposes a WCF service.

Also we have no issue when we load the DocumentDB - just when querying it.

<T> is just Document....   Microsoft.Azure.Document.Documents....

      doc = AzureDocDb.GetMetaDataDoc<Document>(iStateReq.NamedName).Result;
                    AssetMetaData metaData = (dynamic) doc;

Code:
         public static async Task<T> GetMetaDataDoc<T>(string noSqlDbName)
        {
            var oListofMetaAssets = await GetMetaData<T>(noSqlDbName, String.Format(SearchUtility.QueryByMetaDataId));
            return oListofMetaAssets.FirstOrDefault();
        }

public static async Task<List<T>> GetMetaData<T>(string noSqlDbName, string queryString)
        {
            var client = GetDocumentDbClient(noSqlDbName);
            return client == null
                ? null
                : client.CreateDocumentQuery<T>(
                    CosmosSingleton.Instance.GetDocColl(String.Empty, SearchUtility.SearchConfig.DocDbs.DatabaseName,
                        SearchMgrEnty.Meta).UriInfo, String.Format(queryString)).ToList();
            //DocumentCollection collDocDb = await GetOrCreateDocDbCollection(client, noSqlDbName, SearchUtility.SearchConfig.DocDbs.DatabaseName, SearchMgrEnty.Meta);
            //return client.CreateDocumentQuery<T>(collDocDb.SelfLink, String.Format(queryString)).ToList();
        }

Model:
public class AssetMetaData
    {
        [JsonProperty(PropertyName = "id")]
        public string Id { get; set; }

        /// <summary>
        /// Document Type - {Meta}
        /// </summary>
        [JsonProperty]
        public string DocType { get; set; }

        [JsonProperty]
        public long ArtistCount { get; set; }

        [JsonProperty]
        public long CompCount { get; set; }

        [JsonProperty]
        public long ShowCount { get; set; }

        [JsonProperty]
        public long SongCount { get; set; }

        [JsonProperty]
        public string MetaCreatedDateTime { get; set; }

        [JsonProperty]
        public string MetaUpdatedDateTime { get; set; }

        [JsonProperty]
        public string CreatedBy { get; set; }

        [JsonProperty]
        public List<RefreshInfo> ListRefreshInfo { get; set; }

**Comment by ealsur** [comment]
I updated the sample repo with your AssetMetadata and [calling the methods described](https://github.com/ealsur/cloudservicedocdbsample/blob/master/WebRole1/Controllers/HomeController.cs) and it still works. 

There are some warnings regarding the use of `async` in methods that are not actually async, but it compiles and runs.

The webapp uses Net Full Framework, no Net Core, it's a good ol' MVC app.

**Comment by SubyBrue** [comment]
Strange maybe its something with the azure Cloud Service project in our solution then?  As its only breaking when it runs there?

Sent from my iPhone

On Jul 12, 2018, at 5:43 PM, Matias Quaranta <notifications@github.com<mailto:notifications@github.com>> wrote:


I updated the sample repo with your AssetMetadata and calling the methods described<https://github.com/ealsur/cloudservicedocdbsample/blob/master/WebRole1/Controllers/HomeController.cs> and it still works.

There are some warnings regarding the use of async in methods that are not actually async, but it compiles and runs.

The webapp uses Net Full Framework, no Net Core, it's a good ol' MVC app.

**Comment by ealsur** [comment]
@SubyBrue It does sound related, since you can confirm that running it as a standalone Web runs fine.

Some setting in the `ccproj` file that might be affecting some dependency? Like some binding redirect? Did you try creating a new Cloud Service project and point the WebRoles to use the ones in your current Cloud Service project? To see if the same web content with a blank Cloud Service works.

**Comment by SubyBrue** [comment]
@ealsur Yes I have made sure that all the versions are the same and that the binding redirects are setup properly.    Haven't tried creating a new Cloud Service project.

I didn't get a chance to look through your full solution yet (will do so later tonight EST) but two things on yours that I noticed.   

A) Yours is still running Newtonsoft 6.0.8, we have upgraded to 11.0.2  and 
B) We changed all package.configs to PackageReference inside the csproj of the dll library projects and web project.

Was there a reason that you didn't move to Newtonsoft 11.0.2 (or a version past 6.0.8) as one of my fellow developers found an article claiming that Microsoft.Azure.DocumentDB from version 1.15 onward didn't play nice with newer version of Newtonsoft.     Unfortunately, the same webproject we have has dependencies on newer versions of Newtonsoft.

**Comment by ealsur** [comment]
@SubyBrue I left all default packages that came with a Cloud Service project creation, I didn't want to customize anything other than the DocumentDB SDK package.

I updated the sample repo to Newtonsoft 11.0.2 and it still works.

Did you remove the `packages.json` from the project files then if you migrated to PackageReference? 

You mentioned you have multiple library projects, were all of them updated to the DocumentDB SDK? Can you share more details on that scenario?

**Comment by SubyBrue** [comment]
The overall architecture has the Solution that has the CloudService project, two Web Projects which are for the WCF services.   There are then 2 shared underlying repository and infrastructure library projects.   The Infrastructure project has a reference to our SearchLibrary.DLL code where I've pulled the samples I'd displayed above, and where the code is dying when running within the Azure Cloud Service.

When we updated the Search Library project, and the underlying repository and infrastructure library's we used the build in (to VS 15.7.4) Migration tool to convert the Package.Config to PackageReference.

For the two web projects that we converted to PackageReference the built in converter wasn't available so we used NuGet PackageReference Upgrader v1.0.1.vsix  to convert it.... 

I'll look at your project again tonight and see if I notice any variations between the two as per the settings on the projects or the Cloud Service.

**Comment by SubyBrue** [comment]
Trying to open your GIT solution and run it and it keeps complaining about Microsoft.Net.Compilers.props and Microsoft.CodeDom.Providers.DotNetCompilerPlatform.props in the csproj file.... I tried to reset them to ones on my machine and when I do that 1/2 of the references disappear

**Comment by ealsur** [comment]
That's odd, I created it with File > New Project in Visual Studio 2015:
![image](https://user-images.githubusercontent.com/1633401/42728231-95a53a14-878c-11e8-9183-3b4ba77310c6.png)

**Comment by SubyBrue** [comment]
Ah I'm using VS2017... after clearing and reinstalling a few of the references I was able to get it to compile but still can't get it running.    Currently its throwing the following error - just need some time to work through it "No connection could be made because the target machine actively refused it"

**Comment by ealsur** [comment]
The sample is connecting to the Cosmos DB Emulator, change the credentials in Startup.cs and HomeController.cs if you want it to run on another account.
________________________________

**Comment by ealsur** [comment]
@SubyBrue Did you try creating a Cloud Service project from scratch and pointing to the 2 existing WebRole projects?

**Comment by SubyBrue** [comment]
Haven't had a chance - hoping to return looking at this in a few days.

**Comment by ealsur** [comment]
@SubyBrue did you finally found the issue? Can this be closed?

**Comment by SubyBrue** [comment]
Yes - very strange after going through and attempting all the fixes you suggested after we upgraded to VS 2017 15.7.5 it started working for us.    Your code repository always worked (with our Cosmo credentials).   I'm wondering if there was some inconsistency with the way the PackageReference upgrade we did was wonky and solved itself with the 15.7.5 release.... As we had another developer on an earlier release 15.7.2 attempt the same code and hit the failures we did.

Thanks Again ealsur....
